### PR TITLE
UDP support

### PR DESCRIPTION
**Pull request by samuelkarp**
Here's a first draft of UDP support.  Note that it is not yet actually supported by the service, so testing has been manual (with some limited testing in our dev environment).  Once the service supports sending protocol, this will address #2.

The following concerns should be addressed by this PR:
* register with UDP resources available
* receive protocol from ACS
* start containers with the correct indicated protocol
* serialize in our saved state file and update the `Version` as UDP ports are backwards-incompatible

There is a minor change in one of the ACS shapes used for establishing the websocket (`agentVersion` field is renamed to `versionInfo`).  The previous name is actually an error, but fortunately neither the error or the change here is impacting as the shape is flattened while establishing the websocket and that name is ignored.

r? @euank @aaithal

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,6 +77,7 @@ configure them as something other than the defaults.
 |:----------------|:----------------------------|:------------|:--------------|
 | `ECS_CLUSTER`       | clusterName             | The cluster this agent should check into. | default |
 | `ECS_RESERVED_PORTS` | `[22, 80, 5000, 8080]` | An array of ports that should be marked as unavailable for scheduling on this Container Instance. | `[22, 2375, 2376, 51678]` |
+| `ECS_RESERVED_PORTS_UDP` | `[22, 80, 5000, 8080]` | An array of UDP ports that should be marked as unavailable for scheduling on this Container Instance. | `[]` |
```
Comment (euank) [inline README.md]: We can have a better example here; may I suggest `[53, 123]` (DNS, NTP respectively)

**Comment by euank** [comment]
I realized that this actually misses the `submitContainerStateChange` submission of protocol, which will need to be there for udp ports to correctly display in describes.
This is a little more involved because it needs SDK client changes too.

Other than that issue, this looks good, but that issue will need to be addressed before it can be merged.

**Comment by samuelkarp** [comment]
@euank Thanks for letting me know.  I can take care of this when I get back to the office next week.

**Comment by samuelkarp** [comment]
@euank Rebased on the rest of the work in the `dev` branch and addressed your comments.  I haven't end-to-end tested this yet, but I think this is ready for review.

**Comment by euank** [comment]
I gave it another good look and found another missing detail. There's a conversion function for docker port map->portbindings which didn't get updated. See the most recent commit on [my branch](https://github.com/euank/amazon-ecs-agent/commits/UdpProtocol) for what I mean.

I'm not sure if you'd like to pull my commit into this PR or if you'd rather me open one in its stead. Either is fine by me.

With the addition of my above-referenced commit, everything looks good to me.

**Comment by samuelkarp** [comment]
Thanks @euank.  I'll cherry-pick your commit onto my branch and rebase everything so it's able to be merged again.  Expect an update on Tuesday.

**Comment by samuelkarp** [comment]
@euank I've incorporated your change along with a few of my own and tested end-to-end in our dev environment.  Thanks for catching the missing conversion function.

**Comment by aaithal** [comment]
lgtm

**Comment by euank** [comment]
LGTM as well. If you want to rebase (including my change) I don't mind, thought it's not split up badly as is. I do think we should wait on merging it until it will work everywhere. Thanks!

**Comment by samuelkarp** [comment]
Thanks!  Since this is backwards-compatible, I'll just merge now.

**Pull request merged**